rts/stubs.c

```diff
@@ -3,15 +3,15 @@
 
 __attribute__ ((noreturn)) void rts_trap(char *msg, uint32_t len);
 
-/* Stubbery for musl */
+// Stubbery for musl. Functions below are used by some of the musl functions we
+// compile. musl implementations use system calls. We don't want them to be
+// used and we can't provide sensible implementations for them, so we provide
+// stubs below that just trap when called.
 
-// NOTE (osa): This function previously did not return even though it should
-// (return type declared as `int`). Returning `0` here for now but we should
-// probably make sure is never used (maybe trap).
-int wctomb(char *s, wchar_t wc) { *s = '\0'; return 0; }
+int wctomb(char *s, wchar_t wc) { rts_trap("wctomb", 6); }
 
-int fputs(const char *s, void *f) { return 0; }
+int fputs(const char *s, void *f) { rts_trap("fputs", 5); }
```

Comment:
Maybe we should only do this for debug mode? When installing canisters, every byte of "ballast" might cross the limit.

Reply:
> When installing canisters, every byte of "ballast" might cross the limit.

Sorry I couldn't follow this. What happens when installing a canister that can execute this code?

Reply:
It is about the IC install message size that carries the Wasm binary. When there is dead code in there that pulls in more strings the Wasm might pass a limit and won't install any more. Here it is just a few bytes, that is probably okay. It was a general comment. Feel free to ignore.

Reply:
Looking at https://github.com/emscripten-core/emscripten/issues/8525, my suggestion is to just have a `__builtin_trap` here which is supposed to lower to the Wasm `unreachable`.

Reply:
That change (481597c) gives
``` wasm
  (func $wctomb (type 2) (param i32 i32) (result i32)
    unreachable
    unreachable)
  (func $fputs (type 2) (param i32 i32) (result i32)
    unreachable
    unreachable)
  (func $strerror (type 3) (param i32) (result i32)
    unreachable
    unreachable)
```